### PR TITLE
ENH more stable gradient of CrossEntropy

### PR DESCRIPTION
**Pull request by lorentzenchr**
Similar to https://github.com/scikit-learn/scikit-learn/pull/28048.

There is a small runtime cost to pay, but gradient computation is not the main bottleneck of histogram based gradient boosting.

### CODE REVIEW

**Comment by jameslamb** [comment]
Thanks for this! I'll defer to @shiyu1994 and @guolinke to review.

Until then, can you please update this to the latest `master`?

File: src/objective/xentropy_objective.hpp [hunk 1]
```diff
@@ -75,21 +75,54 @@ class CrossEntropy: public ObjectiveFunction {
...
       // compute pointwise gradients and Hessians with implied unit weights
       #pragma omp parallel for num_threads(OMP_NUM_THREADS()) schedule(static)
       for (data_size_t i = 0; i < num_data_; ++i) {
-        const double z = 1.0f / (1.0f + std::exp(-score[i]));
-        gradients[i] = static_cast<score_t>(z - label_[i]);
-        hessians[i] = static_cast<score_t>(z * (1.0f - z));
+        if (score[i] > -37.0) {
+          const double exp_tmp = std::exp(-score[i]);
+          gradients[i] = static_cast<score_t>(((1.0f - label_[i]) - label_[i] * exp_tmp) / (1.0f + exp_tmp));
+          hessians[i] = static_cast<score_t>(exp_tmp / (1 + exp_tmp) * (1 + exp_tmp));
```
Comment (shiyu1994) [inline src/objective/xentropy_objective.hpp:106]: The / followed by a * simply returns `exp_tmp`, which is not the expected hessian.
```suggestion
          hessians[i] = static_cast<score_t>(exp_tmp / ((1 + exp_tmp) * (1 + exp_tmp)));

```

Comment (shiyu1994) [reply src/objective/xentropy_objective.hpp:106]: Is it possible that `hessians[i] = static_cast<score_t>(exp_tmp / (1 + exp_tmp) / (1 + exp_tmp));` could be more numerically stable?

File: src/objective/xentropy_objective.hpp [hunk 2]
```diff
@@ -75,21 +75,54 @@ class CrossEntropy: public ObjectiveFunction {
...
+        } else {
+          const double exp_tmp = std::exp(score[i]);
+          gradients[i] = static_cast<score_t>(exp_tmp - label_[i]);
+          hessians[i] = static_cast<score_t>(exp_tmp);
```
Comment (shiyu1994) [inline src/objective/xentropy_objective.hpp:110]:
```suggestion
          hessians[i] = static_cast<score_t>(exp_tmp / ((1 + exp_tmp) * (1 + exp_tmp)));

```

File: src/objective/xentropy_objective.hpp [hunk 3]
```diff
@@ -75,21 +75,54 @@ class CrossEntropy: public ObjectiveFunction {
...
       // compute pointwise gradients and Hessians with given weights
       #pragma omp parallel for num_threads(OMP_NUM_THREADS()) schedule(static)
       for (data_size_t i = 0; i < num_data_; ++i) {
-        const double z = 1.0f / (1.0f + std::exp(-score[i]));
-        gradients[i] = static_cast<score_t>((z - label_[i]) * weights_[i]);
-        hessians[i] = static_cast<score_t>(z * (1.0f - z) * weights_[i]);
+        if (score[i] > -37.0) {
+          const double exp_tmp = std::exp(-score[i]);
+          gradients[i] = static_cast<score_t>(((1.0f - label_[i]) - label_[i] * exp_tmp) / (1.0f + exp_tmp) * weights_[i]);
+          hessians[i] = static_cast<score_t>(exp_tmp / (1 + exp_tmp) * (1 + exp_tmp) * weights_[i]);
```
Comment (shiyu1994) [inline src/objective/xentropy_objective.hpp:120]: Similar as the case without weights.

File: src/objective/xentropy_objective.hpp [hunk 4]
```diff
@@ -75,21 +75,54 @@ class CrossEntropy: public ObjectiveFunction {
...
+        } else {
+          const double exp_tmp = std::exp(score[i]);
+          gradients[i] = static_cast<score_t>((exp_tmp - label_[i]) * weights_[i]);
+          hessians[i] = static_cast<score_t>(exp_tmp * weights_[i]);
```
Comment (shiyu1994) [inline src/objective/xentropy_objective.hpp:124]: Similar as the case without weights.

**Review by shiyu1994 — COMMENTED** [review]
Thanks for the contribution. I just left a few comments about the correction of hessian computation.

File: src/objective/xentropy_objective.hpp (hunk 1 above)
Comment (lorentzenchr) [reply src/objective/xentropy_objective.hpp:106]: First for all, yes I forgot the parenthesis. Thanks for spotting it. It is surprising that still all the tests pass (with this bug).

Then `(exp_tmp / (1 + exp_tmp) / (1 + exp_tmp))` is more numerical stable in the sense that it could prevent overflow. But `exp_tmp > exp(37) = 1e16` and squaring that is within even single precision (3e38), and note that `exp_tmp` is even double precision.

Comment (shiyu1994) [reply src/objective/xentropy_objective.hpp]: Thanks. Could you also fix the hessian calculation in the else branch?

File: src/objective/xentropy_objective.hpp (hunk 2 above)
Comment (lorentzenchr) [reply src/objective/xentropy_objective.hpp:110]: This is not needed as `exp_tmp < 1e-16` is tiny and `(1 + exp_tmp)` is just 1. Otherwise stated, the implemented formula is the 1st order Taylor series in `exp_tmp`.

Comment (shiyu1994) [reply src/objective/xentropy_objective.hpp:110]: I see. That makes sense.

Comment (shiyu1994) [reply src/objective/xentropy_objective.hpp:110]: But maybe it would still be better to write the original calculation formula explicitly to avoid ambiguity?

**Review by shiyu1994 — APPROVED** [review]

File: src/objective/xentropy_objective.hpp (hunk 2 above)
Comment (lorentzenchr) [reply src/objective/xentropy_objective.hpp:110]: What do you mean with "ambiguity"?
